### PR TITLE
Add support set/list cardinality to lucene

### PR DESCRIPTION
**Pull request by mad**
Fix for #670 

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [ ] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java [hunk 1]
```diff
@@ -330,23 +349,59 @@ private Document retrieveOrCreate(String docID, IndexSearcher searcher) throws I
...
+            if (!isNew) {
+                Iterator<IndexableField> it = doc.iterator();
+                while (it.hasNext()) {
+                    IndexableField field = it.next();
+                    if (field.name().equals(e.field)) {
+                        if (ki.getCardinality() == Cardinality.SINGLE) {
+                            it.remove();
+                            break;
+                        } else if (ki.getCardinality() == Cardinality.SET) {
+                            if (convertToStringValue(e.value).equals(field.stringValue())) {
```
[the remark this answers is not in the document]
Comment (porunov) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java]: Also agree. We can use `&&`.

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java [hunk 2]
```diff
@@ -330,23 +349,59 @@ private Document retrieveOrCreate(String docID, IndexSearcher searcher) throws I
...
             Preconditions.checkArgument(!e.hasMetaData(), "Lucene index does not support indexing meta data: %s", e);
             if (log.isTraceEnabled())
                 log.trace("Adding field [{}] on document [{}]", e.field, doc.get(DOCID));
 
-            if (doc.getField(e.field) != null) {
-                doc.removeFields(e.field);
+            KeyInformation ki = information.get(e.field);
+            if (!isNew) {
+                Iterator<IndexableField> it = doc.iterator();
+                while (it.hasNext()) {
```
Comment (porunov) [inline janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java]: As for me, the logic is too complicated to be in one method. I would suggest to move logic under `if (!isNew)` into a separate method.

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java [hunk 3]
```diff
@@ -242,16 +248,28 @@ private void mutateStores(Map.Entry<String, Map<String, IndexMutation>> stores,
...
                 for (final IndexEntry del : mutation.getDeletions()) {
                     Preconditions.checkArgument(!del.hasMetaData(), "Lucene index does not support indexing meta data: %s", del);
-                    final String key = del.field;
-                    if (doc.getField(key) != null) {
-                        if (log.isTraceEnabled())
-                            log.trace("Removing field [{}] on document [{}]", key, documentId);
-
-                        doc.removeFields(key);
+                    String fieldName = del.field;
+                    if (log.isTraceEnabled())
```
[the remark this answers is not in the document]
Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java]: Done

File: janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java (hunk 2 above)
Comment (mad) [reply janusgraph-lucene/src/main/java/org/janusgraph/diskstorage/lucene/LuceneIndex.java]: Done

File: janusgraph-lucene/src/test/java/org/janusgraph/diskstorage/lucene/BerkeleyLuceneTest.java [hunk 4]
```diff
@@ -69,6 +74,7 @@ public void testPrintSchemaElements() {
 
         String expected = "A18422278042949EE2B162837EC27A63";
         String output = mgmt.printSchema();
+        System.out.println(output);
```
Comment (porunov) [inline janusgraph-lucene/src/test/java/org/janusgraph/diskstorage/lucene/BerkeleyLuceneTest.java]: Looks like this is added by accident

File: janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java [hunk 5]
```diff
@@ -1080,10 +1089,10 @@ public void testScroll() throws BackendException {
     public void testPersistantIndexData() throws BackendException {
         final String store = "vertex";
 
-        final Multimap<String, Object> initialDoc = HashMultimap.create();
+        final Multimap<String, Object> initialDoc = ArrayListMultimap.create();
         initialDoc.put(DATE, Instant.ofEpochSecond(1));
-        double latitude = random.nextDouble() * 180 - 90;
-        double longitude = random.nextDouble() * 360 - 180;
+        double latitude = 30;
+        double longitude = 60;
```
Comment (porunov) [inline janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java]: Would you be able also to add this flaky test fix into `0.3` branch?

Comment (mad) [reply janusgraph-test/src/main/java/org/janusgraph/diskstorage/indexing/IndexProviderTest.java]: Ok, https://github.com/JanusGraph/janusgraph/pull/1696

**Pull request merged**